### PR TITLE
Product stock adjustments with manual order line item changes

### PR DESCRIPTION
**Pull request by mikejolley**
This is an attempt to improve #21754 to try avoiding reverting to #22012.

[Reasoning explained in the issue comment here.](https://github.com/woocommerce/woocommerce/issues/21754#issuecomment-451109478)

This makes 3 main changes:

1. If an item has already reduced stock in an order and is DELETED, the stock for the deleted item is restored. A note is added.

2. If an item is added to an order manually, stock is reduced immediately for that line item and a note is added.

3. Finally, if you edit the qty level of an item which has already had it's stock reduced, it will increase or decrease the difference:

Everything else is the same. Automated stock events during status change will not double reduce anything because the amounts are stored per line item.

This PR could use feedback from the users in #21754 before merge. I'd recommend for 3.6.

### Changelog entry

> When adding, editing, and deleting items manually from orders, the corresponding product's stock will be updated to reflect the event and an order note will be logged as a record of it happening.

### CODE REVIEW

**Comment by ben72** [comment]
I still miss the way to easily highlight lines when picking items for an order. Say you have 20 items to get from different places. It's great if you can see the progress just by clicking the lines you've picked so far. Makes sense?

**Comment by rrennick** [comment]
> I still miss the way to easily highlight lines when picking items for an order.

See https://github.com/woocommerce/woocommerce/issues/21754#issuecomment-451153589

**Comment by mikejolley** [comment]
Having lines that highlight on click will likely cause confusion as to what they are doing or what is happening. @ben72 Sounds like a potential extension, marking items as picked in a way that persists across page loads. Could be line item meta data.

**Comment by mikejolley** [comment]
@windsorraw Ill reply in the issue.

**Comment by willgorham** [comment]
I really like these changes. Great work! FYI I think this PR covers #22107 as well.

**Comment by ben72** [comment]
@mikejolley Sounds reasonable if few people use it like that. I'm just reacting as features I use go away..

**Comment by filpet** [comment]
@mikejolley just wondering, would it possible to maybe include product exchange with this new way of editing order lines?

E.g. customer orders blue socks, shop runs out of blue socks and asks customer if red socks are ok.
Customer approves, now the shop owner has to delete the blue socks and add red socks in the same quantity and so on...would be a lot easier if one could just switch the product.

**Comment by filpet** [comment]
@mikejolley oh and one more thing which disappeared with removing the line selection:
In case we have to delete multiple products from the order, it's quite a task as we have to delete each single line by itself, one can't select multiple rows! Maybe some kind of multiple selection function could be added back again.

**Comment by gpadilha** [comment]
@mikejolley Nice job. Just wondering if the DELETED, on item 1, includes trashed orders or it would be just for permanent deleted orders.

**Comment by fatpotanga** [comment]
Interesting. For what it's worth I too really miss the hilighted line feature. I found it really useful.
 
Would this work for say, lost orders whereby you need to resend in full?
You'd need to mark the stock levels down, but not adjust the order $value

Before the update you could select multiple items and click a single button to increase or reduce stock. This was also useful for exchanges (which I've never found an easy way to do, but that's another conversation.. :)) or having to send out replacement items that were say faulty or damaged.

File: includes/admin/wc-admin-functions.php [hunk 1]
```diff
@@ -178,17 +180,61 @@ function woocommerce_settings_get_option( $option_name, $default = '' ) {
...
+ * Sees if line item stock has already reduced stock, and whether those values need adjusting e.g. after changing item qty.
+ *
+ * @since 3.6.0
+ * @param WC_Order_Item $item Item object.
+ * @param integer       $item_quantity Optional quantity to check against. Read from object if not passed.
+ * @return boolean|array|WP_Error Array of changes or error object when stock is updated (@see wc_update_product_stock). False if nothing changes.
+ */
+function wc_maybe_adjust_line_item_product_stock( $item, $item_quantity = -1 ) {
+	if ( 'line_item' !== $item->get_type() ) {
+		return;
```
Comment (claudiulodro) [inline includes/admin/wc-admin-functions.php]: This should explicitly return something (`false` probably).

File: includes/admin/wc-admin-functions.php [hunk 2]
```diff
@@ -178,17 +180,61 @@ function woocommerce_settings_get_option( $option_name, $default = '' ) {
...
+	}
+
+	$product               = $item->get_product();
+	$item_quantity         = wc_stock_amount( $item_quantity >= 0 ? $item_quantity : $item->get_quantity() );
+	$already_reduced_stock = wc_stock_amount( $item->get_meta( '_reduced_stock', true ) );
+
+	if ( ! $product || ! $product->managing_stock() || ! $already_reduced_stock || $item_quantity === $already_reduced_stock ) {
+		return;
```
Comment (claudiulodro) [inline includes/admin/wc-admin-functions.php]: This should explicitly return something (`false` probably).

**Review by claudiulodro — CHANGES_REQUESTED** [review]
Works great in my testing, and the code looks good. Just a couple areas that need minor tweaks.

The extra order notes about stock level changes are a great idea! That should make it super clear what's going on with item stocks. 👍

**Comment by davidjean170** [comment]
This is exactly what i need. Is this will be integrate in a future update of WooCommerce or in a plugin ?

**Comment by bamarlay** [comment]
This is exactly what I need for my store, and how it should be working. It’s not very practical to manually reduce/ increase stock when products are added or removed from the order. Really looking forward to this feature.

**Comment by claudiosanches** [comment]
@gpadilha 

> Just realized that you meant deleted products from the orders. Should products also be restocked when moving to trash orders?

No, see that moving an order to trash doesn't only means that you want to restock all your products, sometimes is just to remove from your database.
So if you want to restock all products in an order, it's better of doing it inside the order.

@davidjean170

> This is exactly what i need. Is this will be integrate in a future update of WooCommerce or in a plugin ?

Probably in WooCommerce 3.6.

**Review by claudiosanches — APPROVED** [review]
LGTM!

**Comment by davidlonguk** [comment]
Will this be available as a standalone release so we can get this fix without introducing new bugs from 3.6?

**Comment by mikejolley** [comment]
@davidlonguk Nope but feel free to copy paste from the PR and introduce new bugs yourself :p 

FYI this PR doesn't fix bugs, it's new functionality.

**Comment by davidlonguk** [comment]
@mikejolley :) I may well just do that. I see your point about it being new functionality but we are using it to resolve a problem introduced in 3.5 hence reference to bugs. Looks like our choices are to roll back, patch with this new feature or make our own solution so we will try copying your PR as our first choice.

**Comment by tldawson** [comment]
Is there any way to prevent stock change on orders with specific order statuses? I'm reading through the merged commits but I don't see any hooks to change the new functionality. 

For example, if an order had a custom status like `quote` I wouldn't want to change stock until the customer approves the quote (and we change its status to `pending`).

**Comment by Spreeuw** [comment]
+1 on filters for preventing stock change on specific order statuses

**Comment by LeonWesten** [comment]
> Is there any way to prevent stock change on orders with specific order statuses? I'm reading through the merged commits but I don't see any hooks to change the new functionality.
> 
> For example, if an order had a custom status like `quote` I wouldn't want to change stock until the customer approves the quote (and we change its status to `pending`).

Do you know how to fix it? Same problem here..

**Comment by davidlonguk** [comment]
@LeonWesten We are going to be making our plugin that will restore previous behaviour (so there are stock Up/Down buttons ) as we need it for several of our actions that include disabling a line item but not an entire order so we don't have an order status change. If the community would find it useful we can make it publicly available but it's 6-8 weeks out.

**Comment by tldawson** [comment]
Rather than make a plugin, why not create a PR that adds hooks to this code? If I have time this week, I'll see if I can draft up a PR.

**Comment by LeonWesten** [comment]
> Rather than make a plugin, why not create a PR that adds hooks to this code? If I have time this week, I'll see if I can draft up one.

Yes, please..

**Comment by davidlonguk** [comment]
@tldawson That sounds great - I hope it is is accepted but I think that woo have plans to go in a different direction with the stock levels automatically changing when you add/remove line items from an order in a future update rather than going back to manual. We require manual as we have cases where line items remain but we need to increase stock.

**Comment by tldawson** [comment]
I see no reason why they would reject a thoughtful PR that adds hooks to this code. Hooks are the heart of WordPress and Woocommerce's extensibility.

**Comment by LeonWesten** [comment]
Is there already a solution for this problem?

**Comment by AKSUTAH** [comment]
Being able to click and (highlight) multiple items in the order and then increase or decrease stock is essential. Doing them one by one is tedious.
That makes managing orders take way more time and effort.
Sure, keep this feature, but bring the old one back too.

**Comment by AKSUTAH** [comment]
Who do I have to PAY to get the Reduce Stock & Increase Stock Buttons back in the order?
This is getting to the point of extortion. How much do you want to bring this simple necessary feature BACK?
Why was it removed in the first place?

**Comment by AKSUTAH** [comment]
Who do I have to PAY to get the Reduce Stock & Increase Stock Buttons back in the order?
This is getting to the point of extortion. How much do you want to bring this simple necessary feature BACK?
Why was it removed in the first place?

**Pull request merged**